cvmfs/kvstore.cc

```diff
@@ -0,0 +1,233 @@
+  perf::Inc(counters_.n_getbuffer);
+  // LogCvmfs(kLogKvStore, kLogDebug, "get buffer %s", id.ToString().c_str());
+  return entries_.Lookup(id, buf);
+}
+
+bool MemoryKvStore::PopBuffer(const shash::Any &id, MemoryBuffer *buf) {
+  perf::Inc(counters_.n_popbuffer);
+  WriteLockGuard guard(rwlock_);
+  if (!entries_.Lookup(id, buf)) {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on PopBuffer",
+             id.ToString().c_str() );
+    return false;
+  }
+  assert(used_bytes_ >= (*buf).size);
+  used_bytes_ -= (*buf).size;
+  counters_.sz_size->Set(used_bytes_);
+  assert(entry_count_ > 0);
+  --entry_count_;
+  entries_.Forget(id);
+  // LogCvmfs(kLogKvStore, kLogDebug, "popped %s (%uB)", id.ToString().c_str(),
+  //          (*buf).size);
+  return true;
+}
+
+int64_t MemoryKvStore::GetSize(const shash::Any &id) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_getsize);
+  if (entries_.Lookup(id, &mem)) {
+    // LogCvmfs(kLogKvStore, kLogDebug, "%s is %u B", id.ToString().c_str(),
+    //          mem.size);
+    return mem.size;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug,
+             "miss%s on GetSize",
+             id.ToString().c_str());
+    return -ENOENT;
+  }
+}
+
+int64_t MemoryKvStore::GetRefcount(const shash::Any &id) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_getrefcount);
+  if (entries_.Lookup(id, &mem)) {
+    // LogCvmfs(kLogKvStore, kLogDebug, "%s has refcount %u",
+    //          id.ToString().c_str(), mem.refcount);
+    return mem.refcount;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on GetRefcount",
+             id.ToString().c_str());
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::IncRef(const shash::Any &id) {
+  perf::Inc(counters_.n_incref);
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (entries_.Lookup(id, &mem)) {
+    assert(mem.refcount < UINT_MAX);
+    ++mem.refcount;
+    entries_.Insert(id, mem);
+    LogCvmfs(kLogKvStore, kLogDebug, "increased refcount of %s to %u",
+             id.ToString().c_str(), mem.refcount);
+    return true;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on IncRef",
+             id.ToString().c_str());
+    return false;
+  }
+}
+
+bool MemoryKvStore::Unref(const shash::Any &id) {
+  perf::Inc(counters_.n_unref);
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (entries_.Lookup(id, &mem)) {
+    assert(mem.refcount > 0);
+    --mem.refcount;
+    entries_.Insert(id, mem);
+    LogCvmfs(kLogKvStore, kLogDebug, "decreased refcount of %s to %u",
+             id.ToString().c_str(), mem.refcount);
+    return true;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on Unref",
+             id.ToString().c_str());
+    return false;
+  }
+}
+
+int64_t MemoryKvStore::Read(
+  const shash::Any &id,
+  void *buf,
+  size_t size,
+  size_t offset
+) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_read);
+  ReadLockGuard guard(rwlock_);
+  if (entries_.Lookup(id, &mem)) {
```

Comment:
It's probably easier to read to quit early on the error case.  The actual logic can then be un-indented.